File: trading-core/lib/regime.ml
type t = Types.market_regime

(* Classify regime from realized volatility and trend strength *)
let classify ~realized_vol ~trend_strength =
  if realized_vol > 0.30 then Types.Crisis
  else if realized_vol > 0.20 && trend_strength < 0.5 then High_vol_ranging
  else if realized_vol < 0.15 && trend_strength > 1.0 then Low_vol_trending
  else Normal

(* Target allocation percentages by regime *)
type allocation = {
  mean_reversion: float;
  sector_rotation: float;
  calendar_seasonal: float;
  momentum: float;
  market_making: float;
  cash: float;
}

let target_allocation = function
  | Types.Low_vol_trending -> {
      mean_reversion = 0.25; sector_rotation = 0.10; calendar_seasonal = 0.05;
      momentum = 0.40; market_making = 0.15; cash = 0.05 }
  | Normal -> {
      mean_reversion = 0.30; sector_rotation = 0.15; calendar_seasonal = 0.10;
      momentum = 0.25; market_making = 0.15; cash = 0.05 }
  | High_vol_ranging -> {
      mean_reversion = 0.35; sector_rotation = 0.10; calendar_seasonal = 0.10;
      momentum = 0.10; market_making = 0.05; cash = 0.30 }
  | Crisis -> {
      mean_reversion = 0.10; sector_rotation = 0.05; calendar_seasonal = 0.05;
      momentum = 0.05; market_making = 0.05; cash = 0.70 }

let to_string = function
  | Types.Low_vol_trending -> "Low_vol_trending"
  | Normal -> "Normal"
  | High_vol_ranging -> "High_vol_ranging"
  | Crisis -> "Crisis"

File: trading-core-js/trading-core/lib/regime.js
// trading-core-js/trading-core/lib/regime.js
function classify(realized_vol, trend_strength) {
  if (realized_vol > 0.30) {
    return (
      /* Crisis */
      3
    );
  } else if (realized_vol > 0.20 && trend_strength < 0.5) {
    return (
      /* High_vol_ranging */
      2
    );
  } else if (realized_vol < 0.15 && trend_strength > 1.0) {
    return (
      /* Low_vol_trending */
      0
    );
  } else {
    return (
      /* Normal */
      1
    );
  }
}
function target_allocation(param) {
  switch (param) {
    case /* Low_vol_trending */
    0:
      return {
        mean_reversion: 0.25,
        sector_rotation: 0.1,
        calendar_seasonal: 0.05,
        momentum: 0.4,
        market_making: 0.15,
        cash: 0.05
      };
    case /* Normal */
    1:
      return {
        mean_reversion: 0.3,
        sector_rotation: 0.15,
        calendar_seasonal: 0.1,
        momentum: 0.25,
        market_making: 0.15,
        cash: 0.05
      };
    case /* High_vol_ranging */
    2:
      return {
        mean_reversion: 0.35,
        sector_rotation: 0.1,
        calendar_seasonal: 0.1,
        momentum: 0.1,
        market_making: 0.05,
        cash: 0.3
      };
    case /* Crisis */
    3:
      return {
        mean_reversion: 0.1,
        sector_rotation: 0.05,
        calendar_seasonal: 0.05,
        momentum: 0.05,
        market_making: 0.05,
        cash: 0.7
      };
  }
}
function to_string(param) {
  switch (param) {
    case /* Low_vol_trending */
    0:
      return "Low_vol_trending";
    case /* Normal */
    1:
      return "Normal";
    case /* High_vol_ranging */
    2:
      return "High_vol_ranging";
    case /* Crisis */
    3:
      return "Crisis";
  }
}
export {
  classify,
  target_allocation,
  to_string
};
